cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -247,4 +265,21 @@ public WorkflowToken deepCopy() {
     }
     return builder.build();
   }
+
+  private void ensureEnoughMemoryAvailable(String key, Value value) {
+    int totalSizeAfterAdding  = Bytes.toBytes(key).length + Bytes.toBytes(value.toString()).length + currentSizeBytes;
+    if (totalSizeAfterAdding > maxSizeBytes) {
+      throw new IllegalStateException(String.format("Exceeded maximum permitted size of workflow token '%sMB'. " +
+                                                      "Current size is '%sMB'. Please increase the maximum " +
+                                                      "permitted size by setting the parameter '%s' in " +
+                                                      "cdap-site.xml to add more values.",
+                                                    maxSizeBytes / (1024 * 1024),
+                                                    currentSizeBytes / (1024 * 1024),
+                                                    Constants.AppFabric.WORKFLOW_TOKEN_MAX_SIZE_MB));
+    }
+  }
+
+  private void updateSize(String key, Value value) {
+    currentSizeBytes += Bytes.toBytes(key).length + Bytes.toBytes(value.toString()).length;
```

Comment:
I had kept these two methods separate, so I check that enough memory is available as one of the first lines in ``put()`` (line 144), while only update ``currentSizeBytes`` (and now ``bytesLeft``) after the put operation was successful (line 162). However, removed the Bytes conversion and just using ``String.length()``. Let me know if that is ok.